orte/mca/schizo/singularity/schizo_singularity.c

```diff
@@ -78,24 +81,57 @@ static int setup_fork(orte_job_t *jdata, orte_app_context_t *app)
     if (NULL != app->env) {
         pth = opal_path_findv("singularity", X_OK, app->env, NULL);
     }
-    if (NULL == pth) {
+    if (NULL != pth) {
+        /* opal_path_findv returned the absolute path to the Singularity binary,
+         * we want the directory where the binary is. */
+        pth = opal_dirname(pth);
+        opal_output_verbose(1, orte_schizo_base_framework.framework_output,
+                            "%s schizo:singularity: Singularity found from env: %s\n",
+                            ORTE_NAME_PRINT(ORTE_PROC_MY_NAME), pth);
+    } else {
         /* wasn't in the environment - see if it was found somewhere */
         if (0 < strlen(OPAL_SINGULARITY_PATH)) {
             if (0 != strcmp(OPAL_SINGULARITY_PATH, "DEFAULT")) {
                 pth = OPAL_SINGULARITY_PATH;
+                opal_output_verbose(1, orte_schizo_base_framework.framework_output,
+                                     "%s schizo:singularity: using default Singularity from %s\n",
+                                     ORTE_NAME_PRINT(ORTE_PROC_MY_NAME), pth);
+                 /* Update (if possible) the PATH of the app so it can find singularity otherwise
+                    it will likely not find it and create a failure. The default path to singularity
+                    that is set at configuration time may not be in the environment that is passed in
+                    by the user. */
+                 for (i = 0; NULL != app->env[i]; i++) {
```

Comment:
Don't worry about fixing this, but this block of code isn't needed. Passing the "prepend_envar" attribute tells the daemon's odls component to prepend the specified path to the PATH envar of the app's procs prior to executing them.

Reply:
I thought it would be the case as well but it did not work for all use cases. If I recall correctly, it did not work with the default singularity. I will test it again. :)

Reply:
I confirm that I am wrong, I just tested it and my problem at the time was obviously from somewhere else. :)

Reply:
Thanks - much appreciate the contribution!